### PR TITLE
Show variable definition as documentation

### PR DESCRIPTION
**Pull request by temanbrcom**
Resolve documentation request immediately

### CODE REVIEW

**Comment by temanbrcom** [comment]
The duplicated logic in the providers is temporary, it will change with introducing the new version of the analysis result

File: server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/ParagraphCompletion.java [hunk 1]
```diff
@@ -14,40 +14,42 @@
...
-    return PARAGRAPHS.prefix;
-  }
-
-  @NonNull
-  @Override
-  public CompletionItemKind getKind() {
-    return CompletionItemKind.Method;
+  private Function<String, CompletionItem> toParagraphCompletion() {
+    return name -> {
+      CompletionItem item = new CompletionItem(name);
```
Comment (asatklichov) [inline server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/ParagraphCompletion.java:47]: Optional: I guess, is it possible to extract this as a  method, then one "return" can removed (implicit lambda return)

File: server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/KeywordCompletion.java [hunk 2]
```diff
@@ -36,27 +39,24 @@
...
-    return KEYWORDS.prefix;
-  }
-
-  @NonNull
-  @Override
-  public CompletionItemKind getKind() {
-    return CompletionItemKind.Keyword;
+  private Function<String, CompletionItem> toKeywordCompletion() {
+    return it -> {
+      CompletionItem item = new CompletionItem(it);
```
Comment (asatklichov) [inline server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/KeywordCompletion.java:53]: Optional: I guess, is it possible to extract this as a  method, then one "return" can removed (implicit lambda return)

File: server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/SectionCompletion.java [hunk 3]
```diff
@@ -14,41 +14,42 @@
...
-  public String getSortOrderPrefix() {
-    return SECTIONS.prefix;
-  }
-
-  @NonNull
-  @Override
-  public CompletionItemKind getKind() {
-    return CompletionItemKind.Function;
+  private Function<String, CompletionItem> toSectionCompletion() {
+    return name -> {
```
Comment (asatklichov) [inline server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/SectionCompletion.java:46]: Optional: I guess, is it possible to extract this as a  method, then one "return" can removed (implicit lambda return)

File: server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/SnippetCompletion.java [hunk 4]
```diff
@@ -39,37 +45,33 @@
...
+        .collect(toList());
   }
 
-  @NonNull
-  @Override
-  public CompletionItem customize(@NonNull CompletionItem item) {
-    item.setInsertText(snippets.getInformationFor(item.getLabel()));
-    item.setInsertTextFormat(InsertTextFormat.Snippet);
-    return item;
+  private Function<String, CompletionItem> toSnippetCompletions() {
```
Comment (asatklichov) [inline server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/SnippetCompletion.java:57]: same

File: server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/VariableCompletion.java [hunk 5]
```diff
@@ -14,40 +14,53 @@
...
+  private Predicate<Variable> matchNames(@NonNull String token) {
+    return it -> it.getName().regionMatches(true, 0, token, 0, token.length());
   }
 
-  @NonNull
-  @Override
-  public CompletionItemKind getKind() {
-    return CompletionItemKind.Variable;
+  private Function<Variable, CompletionItem> toVariableCompletionItem() {
+    return it -> {
```
Comment (asatklichov) [inline server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/VariableCompletion.java:55]: same

**Review by asatklichov — APPROVED** [review]
🌊

File: server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/VariableCompletion.java [hunk 6]
```diff
@@ -14,40 +14,53 @@
...
+      String name = it.getName();
+      CompletionItem item = new CompletionItem(name);
+      item.setLabel(name);
+      item.setInsertText(name);
+      item.setSortText(VARIABLES.prefix + name);
+      item.setDocumentation(it.getFormattedDisplayLine());
```
Comment (grianbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/VariableCompletion.java:61]: But it shows only that variable description, not the variable tree with parents and children like hover feature.

**Review by grianbrcom — APPROVED** [review]
:elephant:

File: server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/VariableCompletion.java (hunk 6 above)
Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/VariableCompletion.java:61]: Really, I will fix it

File: server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/CompletionUtils.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2020 Broadcom.
```
[the remark this answers is not in the document]
Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/CompletionUtils.java]: I have heard a bunch of opinions on that, so I decided non to touch the year at the moment.

File: server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/ParagraphCompletion.java (hunk 1 above)
Comment (temanbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/ParagraphCompletion.java]: Makes sense

File: server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/DocumentationUtils.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+   *
+   * @param variable - a variable to collect the hierarchy
+   * @return string description
+   */
+  public String collectDescription(Variable variable) {
+    StringBuilder prefix = new StringBuilder();
+    List<String> lines = new ArrayList<>();
+    for (String parentLine : parentsDescriptions(variable.getParent())) {
+      lines.add(prefix + parentLine);
+      prefix.append(PREFIX);
```
Comment (grianbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/service/delegates/completions/DocumentationUtils.java:75]: Here you go. You give in to Sonar.

**Review by grianbrcom — APPROVED** [review]
:broom:

**Pull request merged**